operator/controllers/runtimetask_controller.go

```diff
@@ -18,36 +18,324 @@ package controllers
+		if err != nil {
+			return
+		}
+	}
+	// Handle non-deleted/non recovered Task
+	// NB. in case or task recovered from error, we are forcing another reconcile before actually
+	// executing the next command so the user get evidence of what is happening
+	if !recovered {
+		err = r.reconcileNormal(executionMode, task, log)
+		if err != nil {
+			return
+		}
+	}
+
+	// Always reconcile Task Phase at the end
+	r.reconcilePhase(task)
+
+	return
+}
+
+func (r *RuntimeTaskReconciler) reconcileRecovery(executionMode operatorv1.OperationExecutionMode, task *operatorv1.RuntimeTask, log logr.Logger) bool {
+	// if there are no error, return
+	if task.Status.ErrorReason == nil && task.Status.ErrorMessage == nil {
+		return false
+	}
+
+	// if there is no error recovery strategy, return
+	if task.Spec.GetTypedTaskRecoveryStrategy() == operatorv1.RuntimeTaskRecoveryUnknownStrategy {
+		return false
+	}
+
+	switch task.Spec.GetTypedTaskRecoveryStrategy() {
+	case operatorv1.RuntimeTaskRecoveryRetryingFailedCommandStrategy:
+		log.WithValues("command", task.Status.CurrentCommand).Info("Retrying command after failure")
+		r.recorder.Event(task, corev1.EventTypeNormal, "TaskErrorRetry", fmt.Sprintf("Retrying command %d after failure", task.Status.CurrentCommand))
+	case operatorv1.RuntimeTaskRecoverySkippingFailedCommandStrategy:
+		log.WithValues("command", task.Status.CurrentCommand).Info("Skipping command after failure")
+		r.recorder.Event(task, corev1.EventTypeNormal, "TaskErrorSkip", fmt.Sprintf("Skipping command %d after failure", task.Status.CurrentCommand))
+
+		// Move to the next command
+		task.Status.NextCurrentCommand(task.Spec.Commands)
+		if executionMode == operatorv1.OperationExecutionModeControlled {
+			task.Status.Paused = true
+		}
+
+		// if all the commands are done, set the Task completion time
+		if int(task.Status.CurrentCommand) > len(task.Spec.Commands) {
+			task.Status.SetCompletionTime()
+		}
+	default:
+		//TODO: error (if possible do validation before getting here)
+	}
+
+	// Reset the error
+	task.Status.ResetError()
+
+	// Reset the recovery mode so the user can choose again how to proceed at the next error
+	task.Spec.RecoveryMode = ""
+
+	return true
+}
+
+func (r *RuntimeTaskReconciler) reconcilePauseOverride(operationPaused bool, task *operatorv1.RuntimeTask) {
+	// record paused override state change, if any
+	pausedOverride := operationPaused
+	recordPausedChange(r.recorder, task, task.Status.Paused, pausedOverride, "by top level objects")
+
+	// update status with paused override setting from top level objects
+	task.Status.Paused = pausedOverride
+}
+
+func (r *RuntimeTaskReconciler) reconcileNormal(executionMode operatorv1.OperationExecutionMode, task *operatorv1.RuntimeTask, log logr.Logger) error {
+	// If the Task doesn't have finalizer, add it.
+	//if !util.Contains(task.Finalizers, operatorv1.RuntimeTaskFinalizer) {
+	//	task.Finalizers = append(task.Finalizers, operatorv1.RuntimeTaskFinalizer)
+	//}
```

Comment:
remove commented code or add TODO?

Reply:
Finalizers implementation is still pending il all the controllers, I don't have yet had time to figure out if they should go away or not... Until then I prefer to keep them into my way